src/libcore/cell.rs

```diff
@@ -753,6 +765,39 @@ impl<T> RefCell<T> {
 }
 
 impl<T: ?Sized> RefCell<T> {
+    /// Query the current state of this `RefCell`
+    ///
+    /// The returned value can be dispatched on to determine if a call to
+    /// `borrow` or `borrow_mut` would succeed.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(borrow_state)]
+    ///
+    /// use std::cell::{BorrowState, RefCell};
+    ///
+    /// let c = RefCell::new(5);
+    ///
+    /// match c.borrow_state() {
+    ///     BorrowState::Writing => println!("Cannot be borrowed"),
+    ///     BorrowState::Reading => println!("Cannot be borrowed mutably"),
+    ///     BorrowState::Unused => println!("Can be borrowed (mutably as well)"),
+    /// }
+    /// ```
+    #[unstable(feature = "borrow_state", issue = "27733")]
+    #[inline]
+    pub fn borrow_state(&self) -> BorrowState {
+        let borrow = self.borrow.get();
+        if is_writing(borrow) {
+            BorrowState::Writing
+        } else if is_reading(borrow) {
+            BorrowState::Reading
```

Comment:
The implementation makes sense but it isn't congruent with the documentation above wrt. "would succeed". In particular, getting `Reading` as a result does not mean you may successfully `.borrow()` because `borrow == isize::max_value()` may occur and so `.borrow()` will not be successful. This edge case likely means that panicking on `.borrow()` is just fine, but I think the documentation needs some adjustment nonetheless.

Reply:
Well I don't care about `isize::max_value()`, given I'm not actually going to increase the borrow flag, I don't see why I would want an `will_borrow_succeed` that will return `false` if the borrow flag is `isize::max_value()`.

Reply:
Sure, but even so, the docs need to be updated to document this edge case that `Reading` does not entail `.borrow()` will always succeed.

Reply:
I tweaked the documentation to not mention `borrow` and `borrow_mut` at all.